### PR TITLE
Core/Porting OMP for loops to Parallel Utilities - 13

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
This PR ports the OMP Parallel Loops to Parallel Utilities

- ParallelUtilities

**Changelog**
-Replaced Simple OMP For loop with Parallel Utilities - IndexPartition()
-Replaced OMP For loops with Private variables with Parallel Utilities - IndexPartition()
-Replaced OMP for loops with Reduction with Parallel Utilities - SumReduction

### CODE REVIEW

**Review by loumalouomega — APPROVED** [review]
Ok from my side

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Build is failing

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 1]
```diff
@@ -448,50 +444,45 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
+                    break;
                 }
+            }
 
-                if(empty) {
-                    rA(k, k) = BaseType::mScaleFactor;
-                    rb[k] = 0.0;
-                }
+            if(empty) {
+                rA(Index, Index) = BaseType::mScaleFactor;
```
Comment (ashishdarekar) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: @philbucher 
I guess error is here, Is it not allowed ??? 
And, error is only for CentOS, Why is that a case, I am not able to understand it :(

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 2]
```diff
@@ -904,10 +902,9 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
 
             // Count the row sizes
             std::size_t nnz = 0;
-            #pragma omp parallel for reduction(+:nnz)
-            for (int i = 0; i < static_cast<int>(slave_size); ++i) {
-                nnz += indices[BaseType::mSlaveIds[i]].size();
-            }
+            nnz = IndexPartition<std::size_t>(slave_size).for_each<SumReduction<std::size_t>>([&](std::size_t Index){
+                return indices[BaseType::mSlaveIds[Index]].size();
```
Comment (ashishdarekar) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: @philbucher same ...

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 3]
```diff
@@ -379,24 +377,20 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
         const auto it_dof_begin = BaseType::mDofSet.begin();
 
         //NOTE: dofs are assumed to be numbered consecutively in the BlockBuilderAndSolver
-        #pragma omp parallel for
-        for (int k = 0; k<ndofs; k++) {
-            auto it_dof =  it_dof_begin + k;
-
+        IndexPartition<std::size_t>(ndofs).for_each([&](std::size_t Index){
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: wouldn't it be better to use the block version here?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: this should not happen, and it is definitely not threadsafe (it implies reshaping the CSR matrix).

@loumalouomega can u comment?

**Comment by loumalouomega** [comment]
@RiccardoRossi the csr is reshaped in another part of the code

**Comment by RiccardoRossi** [comment]
what is that "empty" doing?

**Comment by loumalouomega** [comment]
> what is that "empty" doing?

The comment ten lines before: // Detect if there is a line of all zeros and set the diagonal to a 1 if this happens

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 4]
```diff
@@ -281,10 +281,9 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
         BaseType::BuildAndSolve(pScheme, rModelPart, rA, rDx, rb);
 
         // Update the Lagrange multiplier solution
-        #pragma omp parallel for
-        for (int i = 0; i < static_cast<int>(BaseType::mSlaveIds.size()); ++i) {
-            mLagrangeMultiplierVector[i] += rDx[BaseType::mEquationSystemSize + i];
-        }
+        IndexPartition<unsigned int>(static_cast<int>(BaseType::mSlaveIds.size())).for_each([&](unsigned int Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 5]
```diff
@@ -318,10 +317,9 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
         BaseType::BuildRHSAndSolve(pScheme, rModelPart, rA, rDx, rb);
 
         // Update the Lagrange multiplier solution
-        #pragma omp parallel for
-        for (int i = 0; i < static_cast<int>(BaseType::mSlaveIds.size()); ++i) {
-            mLagrangeMultiplierVector[i] += rDx[BaseType::mEquationSystemSize + i];
-        }
+        IndexPartition<unsigned int >(static_cast<int>(BaseType::mSlaveIds.size())).for_each([&](unsigned int Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: same

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 3 above)
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: I agree

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 6]
```diff
@@ -425,20 +419,22 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
             }
-        }
+        });
+
         // Filling with ones the LM dofs
-        #pragma omp parallel for firstprivate(ndofs)
-        for (int k = ndofs; k<static_cast<int>(system_size); ++k) {
-            scaling_factors[k] = 1.0;
-        }
+        int loop_size = static_cast<int>(system_size) - ndofs;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 7]
```diff
@@ -425,20 +419,22 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
+
+        IndexPartition<unsigned int>(loop_size).for_each([&](unsigned int Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 1 above)
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: @ashishdarekar I guess you need to capture it

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 8]
```diff
@@ -448,50 +444,45 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
             }
-        }
+        });
 
-        #pragma omp parallel for firstprivate(system_size)
-        for (int k = 0; k < static_cast<int>(system_size); ++k) {
-            std::size_t col_begin = Arow_indices[k];
-            std::size_t col_end = Arow_indices[k+1];
-            const double k_factor = scaling_factors[k];
+        IndexPartition<int >(static_cast<int>(system_size)).for_each([&](int Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 9]
```diff
@@ -537,15 +528,19 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
-            for (int i = 0; i < static_cast<int>(BaseType::mEquationSystemSize); ++i) {
-                b_modified[i] = rb[i];
-            }
+            int loop_size = static_cast<int>(BaseType::mEquationSystemSize);
+            IndexPartition<std::size_t>(loop_size).for_each([&](std::size_t Index){
+                b_modified[Index] = rb[Index];
+            });
+
+            loop_size = static_cast<int>(total_size_of_system) - static_cast<int>(BaseType::mEquationSystemSize);
+            int start_index = static_cast<int>(BaseType::mEquationSystemSize);
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 10]
```diff
@@ -601,15 +596,18 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
             TSystemVectorType b_modified(total_size_of_system);
 
             // Copy the RHS
-            #pragma omp parallel for
-            for (int i = 0; i < static_cast<int>(BaseType::mEquationSystemSize); ++i) {
-                b_modified[i] = rb[i];
-            }
+            int loop_size = BaseType::mEquationSystemSize;
+            IndexPartition<std::size_t>(loop_size).for_each([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
            IndexPartition<std::size_t>(BaseType::mEquationSystemSize).for_each([&](std::size_t Index){
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 11]
```diff
@@ -601,15 +596,18 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
+
+            loop_size = static_cast<int>(total_size_of_system) - static_cast<int>(BaseType::mEquationSystemSize);
+            int start_index = static_cast<int>(BaseType::mEquationSystemSize);
+
             // Fill with zeros
-            #pragma omp parallel for
-            for (int i = static_cast<int>(BaseType::mEquationSystemSize); i < static_cast<int>(total_size_of_system); ++i) {
-                b_modified[i] = 0.0;
-            }
+            IndexPartition<unsigned int>(loop_size).for_each([&](unsigned int Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: size_t

**Review by philbucher — COMMENTED** [review]
overall looks good, but please use `std::size_t` in general

After those changes are applied we can take a look at the CenOS build

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 4 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: Actually I used this before, but it was showing some error , related to this line (on 284)`BaseType::mEquationSystemSize + Index`, addition is not possible ...If I am not wrong. I will try again and will let you know

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 12]
```diff
@@ -448,50 +444,45 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
             }
-        }
+        });
 
-        #pragma omp parallel for firstprivate(system_size)
-        for (int k = 0; k < static_cast<int>(system_size); ++k) {
-            std::size_t col_begin = Arow_indices[k];
-            std::size_t col_end = Arow_indices[k+1];
-            const double k_factor = scaling_factors[k];
+        IndexPartition<std::size_t>(static_cast<int>(system_size)).for_each([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
        IndexPartition<std::size_t>(system_size).for_each([&](std::size_t Index){
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 13]
```diff
@@ -448,50 +444,45 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
+            std::size_t col_begin = Arow_indices[Index];
+            std::size_t col_end = Arow_indices[Index+1];
+            const double k_factor = scaling_factors[Index];
             if (k_factor == 0.0) {
                 // Zero out the whole row, except the diagonal
                 for (std::size_t j = col_begin; j < col_end; ++j)
-                    if (static_cast<int>(Acol_indices[j]) != k )
+                    if (static_cast<int>(Acol_indices[j]) != Index )
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
                    if (Acol_indices[j] != Index)
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 3 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: Sure, I thought this before, but didn't know about the container available for DOFs ..now I found it. 
I am using block_for_each() with container `Dof<double>`. 
Will make those changes now. 
Thanks @RiccardoRossi & @philbucher

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 4 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: @philbucher, this was the error (in CI)
image

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 1 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: separately ?? `[&]` this includes everything right ?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 4 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: @philbucher Ignore above thing. 
Solved now :)

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 14]
```diff
@@ -281,10 +281,9 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
         BaseType::BuildAndSolve(pScheme, rModelPart, rA, rDx, rb);
 
         // Update the Lagrange multiplier solution
-        #pragma omp parallel for
-        for (int i = 0; i < static_cast<int>(BaseType::mSlaveIds.size()); ++i) {
-            mLagrangeMultiplierVector[i] += rDx[BaseType::mEquationSystemSize + i];
-        }
+        IndexPartition<std::size_t>(static_cast<int>(BaseType::mSlaveIds.size())).for_each([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
        IndexPartition<std::size_t>(BaseType::mSlaveIds.size()).for_each([&](std::size_t Index){
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 15]
```diff
@@ -318,10 +317,9 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
         BaseType::BuildRHSAndSolve(pScheme, rModelPart, rA, rDx, rb);
 
         // Update the Lagrange multiplier solution
-        #pragma omp parallel for
-        for (int i = 0; i < static_cast<int>(BaseType::mSlaveIds.size()); ++i) {
-            mLagrangeMultiplierVector[i] += rDx[BaseType::mEquationSystemSize + i];
-        }
+        IndexPartition<std::size_t>(static_cast<int>(BaseType::mSlaveIds.size())).for_each([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
        IndexPartition<std::size_t>(BaseType::mSlaveIds.size()).for_each([&](std::size_t Index){
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h [hunk 16]
```diff
@@ -373,30 +371,23 @@ class ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier
...
             }
-        }
+        });
 
         // NOTE: The constraints reactions are already computed when solving the dofs
         const int number_slave_dofs = BaseType::mSlaveIds.size();
-        #pragma omp parallel for
-        for (int k = 0; k<number_slave_dofs; k++) {
-            const IndexType equation_id = BaseType::mSlaveIds[k];
+        IndexPartition<std::size_t>(number_slave_dofs).for_each([&](std::size_t Index){
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]:
```suggestion
        IndexPartition<std::size_t>(aseType::mSlaveIds.size()).for_each([&](std::size_t Index){
```

**Review by philbucher — COMMENTED** [review]
there are still quite some usage of int instead of size_t

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h (hunk 1 above)
Comment (ashishdarekar) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h]: Solved. 
Thanks :)

**Comment by loumalouomega** [comment]
I have merged with the master the missing OMP loops migration PR, to ensure no conflict

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
Removing my blocks

**Comment by loumalouomega** [comment]
This is ready for merge

**Comment by philbucher** [comment]
I want to do a thorough review before merging as we recently broke sth

**Review by philbucher — APPROVED** [review]

**Pull request merged**